Replace old value after `null` merges in nested properties when batching merges/updates

fabioh8010 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

https://github.com/Expensify/react-native-onyx/issues/615#issuecomment-2681745461

This PR:

- Changes `Onyx.merge()`/`Onyx.update()` batching logic to replace the current value of a nested property after a `null` change in it, so the subsequent updates of that batch in this nested property can fully reset its data.
- Introduces `OnyxMerge` file that will handle the merging of data in different ways depending on the platform
  - For native, we use `mergeItem` that will take advantage of `JSON_PATCH` and `JSON_REPLACE` SQL operations to merge the object in a performant way. `JSON_PATCH` operations were already being used before, and now we also use `JSON_REPLACE` operations to fully replace the marked nested objects when merging data.
  - For web, we use `setItem` since the object was already merged with its changes before.
- Several changes to `fastMerge`/`mergeObject` in order to mark the nested objects we need to fully replace, more details are commented in the functions.
- Removes `OnyxUtils.removeNullValues` since it won't be necessary anymore after the changes in this PR.
- Creates additional `OnyxUtils.mergeChanges`/`OnyxUtils.mergeAndMarkChanges` to handle the batching of merge changes or just merge data into a value.
- Several unit tests, comments added and types improved.

### Related Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is related to -->

https://github.com/Expensify/react-native-onyx/issues/615

### Automated Tests
<!---
Most changes to Onyx should have accompanying tests. Describe the tests you added or if no tests were added an explanation about why one was not needed.
--->

Unit tests were added to cover these changes.

### Manual Tests
<!---
Each set of changes should be tested against the Expensify/App repo on all platforms to verify it does not break our staging App.
--->

Same as https://github.com/Expensify/App/issues/55199.

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/74347569-094e-4519-a13d-51fb11315882




</details>

<details>
<summary>Android: mWeb Chrome</summary>

I'm having problems with my emulators when opening the Chrome app (they crash instantly), so I couldn't record videos for this platform.

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/b79093c2-78a3-4333-a8e8-08cfa2be3a4c







</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/10a313ed-ddfc-47b9-94fa-dcea950e4ea0


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/e3b389be-01ee-45f8-bc61-fa81b7274204


https://github.com/user-attachments/assets/872d0958-ab04-433f-b59f-458cb49d2309




</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/392971f8-0754-4cf7-83b6-40493ea7be81




</details>

fabioh8010 commented:
Strangely the `should replace the old value after a null merge in a nested property when batching updates` and `should replace the old value after a null merge in a nested property when batching merges` tests are partially passing, which I was not expecting to happen.

Maybe the `fastMerge`/`mergeObject` original logic are partially working somehow? I need to investigate.

fabioh8010 commented:
My tests were "wrong", now it will output the errors in the two scenarios of both Onyx.update() and Onyx.merge()

fabioh8010 commented:
cc @chrispader @neil-marcellini

lakchote commented on lib/Onyx.ts:
```diff
@@ -307,8 +307,10 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
         try {
-            // We first only merge the changes, so we can provide these to the native implementation (SQLite uses only delta changes in "JSON_PATCH" to merge)
-            // We don't want to remove null values from the "batchedDeltaChanges", because SQLite uses them to remove keys from storage natively.
+            // We first only merge the changes, so we can use our custom merging strategy by signalling OnyxUtils.applyMerge()
```
```suggestion
            // We first only merge the changes, so we can use our custom merging strategy by signaling OnyxUtils.applyMerge()
```

lakchote commented on lib/storage/providers/SQLiteProvider.ts:
```diff
@@ -60,28 +61,51 @@ const provider: StorageProvider = {
-            const value = JSON.stringify(pair[1]);
-            return [pair[0], value];
-        });
+        // eslint-disable-next-line @typescript-eslint/prefer-for-of
```
Why did you add this rule disabling (that is repeated through the PR)?

Couldn't you do something like `for (const pair of pairs) {...}`?

fabioh8010 replied:
Because we have another rule that forbids use of `for-of`

> iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations. [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax)

We have other places in the codebase that are using simple for loops like this one, so I suggest we keep the same for now.

lakchote commented on lib/storage/providers/SQLiteProvider.ts:
```diff
@@ -60,28 +61,51 @@ const provider: StorageProvider = {
+            // multiGet() is not guaranteed to return the data in the same order we asked with "nonNullishPairsKeys",
+            // so we use a map to associate keys to their existing values correctly.
+            const existingMap = new Map<OnyxKey, OnyxValue<OnyxKey>>();
+            // eslint-disable-next-line @typescript-eslint/prefer-for-of
```
Same comment as above related to disabling this rule.

fabioh8010 replied:
Answered above

lakchote commented on lib/storage/providers/SQLiteProvider.ts:
```diff
@@ -60,28 +61,51 @@ const provider: StorageProvider = {
+
+            const newPairs: KeyValuePairList = [];
+
+            // eslint-disable-next-line @typescript-eslint/prefer-for-of
```
Same comment as above related to disabling this rule.

fabioh8010 replied:
Answered above

lakchote commented on lib/utils.ts:
```diff
@@ -78,7 +90,28 @@ function mergeObject<TObject extends Record<string, unknown>>(target: TObject |
+                // To achieve this, we first mark these nested objects with an internal flag. With the desired objects
+                // marked, when calling this method again with "shouldReplaceMarkedObjects" set to true we can proceed
+                // to effectively replace them in the next condition.
+                if (isBatchingMergeChanges && targetValue === null) {
```
Could we do something like this?

```suggestion
                if (isBatchingMergeChanges && !targetValue) {
```

lakchote commented on lib/utils.ts:
```diff
@@ -78,7 +90,28 @@ function mergeObject<TObject extends Record<string, unknown>>(target: TObject |
+                // Then, when merging the batched changes with the Onyx value, if a nested object of the batched changes
+                // has the internal flag set, we replace the entire destination object with the source one and remove
+                // the flag.
+                if (shouldReplaceMarkedObjects && sourceValue[ONYX_INTERNALS__REPLACE_OBJECT_MARK] === true) {
```
Could we do something like this?

```suggestion
                if (shouldReplaceMarkedObjects && sourceValue[ONYX_INTERNALS__REPLACE_OBJECT_MARK]) {
```

lakchote reviewed (commented):
Left some comments

chrispader commented on tests/unit/fastMergeTest.ts:
```diff
@@ -86,14 +86,14 @@ describe('fastMerge', () => {
+        const result = utils.fastMerge([1, 2, 3] as any, testObject, true, false, false);
 
         expect(result).toEqual(testObject);
     });
```
Basically test for the extra two parameters that were added

neil-marcellini commented on tests/unit/onyxTest.ts:
```diff
@@ -1698,7 +1698,386 @@ describe('Onyx', () => {
+                    },
+                },
+                [`${ONYX_KEYS.COLLECTION.TEST_UPDATE}entry2`]: {
+                    sub_entry2: {
+                        id: 'sub_entry2',
+                        someKey: 'someValue',
+                        someNestedObject: {
+                            someNestedKey: 'someNestedValue',
+                            anotherNestedObject: {
+                                anotherNestedKey: 'anotherNestedValue',
+                            },
+                            anotherNestedObject2: {
+                                anotherNestedKey2: 'anotherNestedValue2',
+                            },
```
NAB: What case or cases exactly are we testing here? It's kind of chaotic. I would prefer multiple tests of specific cases so it's a bit easier to understand.

fabioh8010 replied:
I've redesigned and simplified all the tests following your suggestion, please let me know later wdyt!

neil-marcellini commented on tests/unit/onyxTest.ts:
```diff
@@ -1698,7 +1698,386 @@ describe('Onyx', () => {
+                                anotherNestedObject: {
+                                    newNestedKey: 'newNestedValue',
+                                },
+                                // Removing "anotherNestedObject3" property in this update.
+                                // This property was only introduced in a previous update, so we don't need to care
+                                // about an old value because there isn't one.
```
I find it really hard to follow these tests. What would make it easier for me is if we appended updates to the updates array one at a time while setting up the test, and as we do so update an `expectedValue` variable. I think that would make it much more clear.

Otherwise I need to break out a pen and paper to track how all the changes should happen in the test 😅 📝

fabioh8010 replied:
I've redesigned and simplified all the tests following your suggestion, please let me know later wdyt!

neil-marcellini commented on lib/storage/providers/SQLiteProvider.ts:
```diff
@@ -60,28 +61,51 @@ const provider: StorageProvider = {
         }
 
-        return this.multiMerge([[key, deltaChanges]]) as Promise<BatchQueryResult>;
+        return this.multiGet(nonNullishPairsKeys).then((storagePairs) => {
+            // multiGet() is not guaranteed to return the data in the same order we asked with "nonNullishPairsKeys",
+            // so we use a map to associate keys to their existing values correctly.
+            const existingMap = new Map<OnyxKey, OnyxValue<OnyxKey>>();
+            // eslint-disable-next-line @typescript-eslint/prefer-for-of
+            for (let i = 0; i < storagePairs.length; i++) {
+                existingMap.set(storagePairs[i][0], storagePairs[i][1]);
+            }
+
+            const newPairs: KeyValuePairList = [];
+
+            // eslint-disable-next-line @typescript-eslint/prefer-for-of
+            for (let i = 0; i < nonNullishPairs.length; i++) {
+                const key = nonNullishPairs[i][0];
+                const newValue = nonNullishPairs[i][1];
+
+                const existingValue = existingMap.get(key) ?? {};
+
+                const mergedValue = utils.fastMerge(existingValue, newValue, true, false, true);
+
+                newPairs.push([key, mergedValue]);
+            }
+
+            return this.multiSet(newPairs);
+        });
```
Yes I think it would be a good idea to measure performance in the app before and after this change to make sure it doesn't cause a large performance hit.

marcaaron replied:
+1 but I am less convinced that this is "fine for now". It feels like we are undoing work that we had a good reason to do at some point. I trust that we are moving in a good direction, but would rather let some benchmarks do the talking.

marcaaron commented:
>now we have some customizations (this very PR) that is impossible to replicate with JSON_PATCH

Can you please update the description to mention which customizations are not compatible with JSON_PATCH? I will review today, thanks!

marcaaron commented:
Is it possible to reword some of this for clarity?

>Changes our merging strategy (defined in applyMerge/fastMerge/mergeObject) to have two additional flags: isBatchingMergeChanges and shouldReplaceMarkedObjects. The isBatchingMergeChanges flag will be used when batching any queued merge changes, as it has a special logic to handle the replacement of old values after null merges. 

Specifically, "handle the replacement of old values after null merges" didn't quite make sense to me.

>The shouldReplaceMarkedObjects flag will be used when applying this batched merge changes to the existing value, using this special logic to effectively replace the old values we desire.

This part of the description is also a bit confusing. Maybe you can also explain what a "marked object" is in the description? Thanks!

marcaaron commented on lib/Onyx.ts:
```diff
@@ -307,8 +307,10 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
         }
 
         try {
-            // We first only merge the changes, so we can provide these to the native implementation (SQLite uses only delta changes in "JSON_PATCH" to merge)
```
A part of me is confused about why we are getting rid of JSON_PATCH. Another part believes that this was a _benefit_ since when we do this we end up passing a smaller delta change instead of a potentially large object and therefore it should be more efficient (hand wave there as I do not entirely recall anymore).

marcaaron commented on lib/Onyx.ts:
```diff
@@ -320,23 +322,23 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
                 return Promise.resolve();
             }
-            const batchedDeltaChanges = OnyxUtils.applyMerge(undefined, validChanges, false);
+            const batchedDeltaChanges = OnyxUtils.applyMerge(undefined, validChanges, false, true, false);
```
NAB, `applyMerge()` is getting a bit complex with all the boolean flags. Curious if there is a way to maybe create a new function instead that does what we want and improve the clarity?

fabioh8010 replied:
I've decided to simplify Simplify and extract `applyMerge()` batching merges logic to a separate `batchMergeChanges()` function which will take care of this part specifically.

marcaaron commented on lib/Onyx.ts:
```diff
@@ -320,23 +322,23 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
             // Case (1): When there is no existing value in storage, we want to set the value instead of merge it.
             // Case (2): The presence of a top-level `null` in the merge queue instructs us to drop the whole existing value.
-            // In this case, we can't simply merge the batched changes with the existing value, because then the null in the merge queue would have no effect
+            // In this case, we can't simply merge the batched changes with the existing value, because then the null in the merge queue would have no effect.
```
I wonder if this comment is making sense still or perhaps just requires clarification? Has "simply merge the batched changes with the existing value" lost it's meaning now that our "merge" has a different behavior?

fabioh8010 replied:
The comment still makes sense because our merge strategy stays the same regarding what is being described there.

marcaaron commented on lib/Onyx.ts:
```diff
@@ -346,11 +348,12 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
                 return Promise.resolve();
             }
 
-            // For providers that can't handle delta changes, we need to merge the batched changes with the existing value beforehand.
```
flagging this since I still have questions about `JSON_PATCH` before we remove it...

marcaaron commented on lib/Onyx.ts:
```diff
@@ -346,11 +348,12 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
-            // For providers that can't handle delta changes, we need to merge the batched changes with the existing value beforehand.
-            // The "preMergedValue" will be directly "set" in storage instead of being merged
-            // Therefore we merge the batched changes with the existing value to get the final merged value that will be stored.
+            // The "preMergedValue" will be directly "set" in storage instead of being merged, so we merge
```
**thinking out loud:** we should update this to mention why we have a "preMergedValue" at all and why it will be "set". Without that explanation that follows is very confusing. Consider:

- variable is named "preMergedValue" implying that it's a step that will happen before a merge.
- we say it will be "set" instead of "merged"
- we then say we will "merge" the batched changes to get the final "merged" value

Maybe we can use more specific language for these things?

fabioh8010 replied:
I rephrased the comments, please let me know if it's ok now!

marcaaron commented on lib/Onyx.ts:
```diff
@@ -346,11 +348,12 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
             // We can remove null values from the "preMergedValue", because "null" implicates that the user wants to remove a value from storage.
-            const preMergedValue = OnyxUtils.applyMerge(shouldSetValue ? undefined : existingValue, [batchedDeltaChanges], true);
+            // Additionally, we will signal OnyxUtils.applyMerge() to replace any nested properties previously marked in "batchedDeltaChanges"
+            // by our custom merging strategy.
```
I agree. I would love if the idea of "marking" could be broken down or documented somewhere if it's not already. Both "marked" and "pre merge" are a bit difficult to grasp without rolling your sleeves up 😄

fabioh8010 replied:
Agree, I just removed this comment and will add more explanations to the places where we actually have the merging logic

marcaaron commented on lib/Onyx.ts:
```diff
@@ -365,7 +368,7 @@ function merge<TKey extends OnyxKey>(key: TKey, changes: OnyxMergeInput<TKey>):
             }
 
-            return Storage.mergeItem(key, batchedDeltaChanges as OnyxValue<TKey>, preMergedValue as OnyxValue<TKey>, shouldSetValue).then(() => {
+            return Storage.mergeItem(key, preMergedValue as OnyxValue<TKey>).then(() => {
```
I think maybe `preMergedValue` should be `batchValue`. That seems more clear to me. Like yes, we still have another "merge" to perform. But since we do a "merge" to get to that value it is confusing to refer to it as "pre merge". Just my 2 cents.

fabioh8010 replied:
I decided to change to `resultValue`, I think it conveys more the meaning of the variable, wdyt?

marcaaron commented on lib/OnyxUtils.ts:
```diff
@@ -1259,6 +1259,8 @@ function applyMerge<TValue extends OnyxInput<OnyxKey> | undefined, TChange exten
     changes: TChange[],
     shouldRemoveNestedNulls: boolean,
+    isBatchingMergeChanges: boolean,
+    shouldReplaceMarkedObjects: boolean,
```
Add docs for these ?

fabioh8010 replied:
Not necessary anymore as I simplified the function.
https://github.com/Expensify/react-native-onyx/pull/620#discussion_r2041183537

marcaaron reviewed (changes requested):
This is generally looking good. Thanks for the work on it @fabioh8010 @chrispader. My main feedback (that @neil-marcellini already called out) would be to make sure that `JSON_PATCH` in the native provider is something we can confidently get rid of. If so, then 👍.

neil-marcellini commented:
@fabioh8010 please request our reviews again when it's ready

fabioh8010 commented:
> @fabioh8010 please request our reviews again when it's ready

Sure, I started working through the comments today and preparing better explanation for you before requesting.

neil-marcellini commented:
@fabioh8010 what is the latest update? How is it coming along?

fabioh8010 commented:
# Performance Analysis

For these analysis I tested on Android (the only provider that have substantial changes is the native one) and profiled `Storage.multiMerge` and `Storage.mergeItem` using the [Performance API](https://developer.mozilla.org/en-US/docs/Web/API/Performance).

## Baseline (main)

| Test | Method | Duration (ms) |
| - | - | - |
| Merging 1000 records with `mergeCollection()` | Storage.multiMerge | 83 |
| Merging 1 record with `mergeCollection()` | Storage.multiMerge | 4 |
| Merging the same record 10 times with `merge()` | Storage.mergeItem (uses Storage.multiMerge under the hood) | 3 |

## Delta (this PR)

| Test | Method | Duration (ms) |
| - | - | - |
| Merging 1000 records with `mergeCollection()` | Storage.multiMerge | 118 (+42,17%) |
| Merging 1 record with `mergeCollection()` | Storage.multiMerge | 37 (+89,19%) |
| Merging the same record 10 times with `merge()` | Storage.mergeItem (uses Storage.setItem under the hood) | 30 (+86,67%) |

---

As you can see the operations are quite more expensive now, probably due to having to use `this.multiGet()` and then `this.multiSet()` instead of JSON_PATCH.

I will try to look for another solution for the native provider, maybe I can keep the JSON_PATCH and only do additional SQL operations for the properties we need to fully replace.

fabioh8010 commented:
**Updates:**

- Investigating and experimenting a way to keep JSON_PATCH operation and then execute the object replacements with separate SQL operations
- Also investigating why a simple `setItem` operation is taking whole 30 ms even on `main`, while `mergeItem` takes around 3-4 ms on `main`. I feel an SQL "SET" operation should have the same or even less execution time than a JSON_PATCH one.

fabioh8010 commented:
Just FYI that I will be OOO next week and will return on May 5th 🌴

fabioh8010 commented:
**Updates:**

- Resuming work on this one this week.

fabioh8010 commented:
**Updates:**

- Designing/working on the `JSON_PATCH` + separate `JSON_REPLACE` operations solution.

fabioh8010 commented:
**Updates:**

- I think I have a working solution for `JSON_PATCH` + separate `JSON_REPLACE` operations approach, will do a cleanup, more testing and finally the benchmarks.

chrispader commented:
> **Updates:**
> 
> * I think I have a working solution for `JSON_PATCH` + separate `JSON_REPLACE` operations approach, will do a cleanup, more testing and finally the benchmarks.

Cool! Lmk if you need any help with the Onyx storage layer or SQLite related stuff, since i'm the maintainer of `react-native-nitro-sqlite` i might be able to help there! 🙌🏼

fabioh8010 commented:
> > **Updates:**
> > 
> > * I think I have a working solution for `JSON_PATCH` + separate `JSON_REPLACE` operations approach, will do a cleanup, more testing and finally the benchmarks.
> 
> Cool! Lmk if you need any help with the Onyx storage layer or SQLite related stuff, since i'm the maintainer of `react-native-nitro-sqlite` i might be able to help there! 🙌🏼

For sure man, thanks! I need to do some cleanup before, and for now I won't merge with `main` yet since I don't want to pollute the benchmarks.

fabioh8010 commented:
**Updates:**

- Did the benchmarks and the logic I created to generate the JSON_REPLACE operations are causing some performance impact when merging big collections (like the `Merging 1000 records with mergeCollection()` case) because we need to traverse the objects again to search for the markers.
- To improve that I passed this logic to `mergeObject` because we can reuse the traverse object logic that already happen there to remove the performance impact.
- It seems to be working but needs more testing. Also started some discussion with @chrispader in DM.

fabioh8010 commented:
**Updates:**

- Merged Chris' changes into my [JSON_PATCH solution branch](https://github.com/Expensify/react-native-onyx/compare/main...callstack-internal:react-native-onyx:bugfix/615-solution2-jsonPatchReplace).
- Fixing some TS errors and doing cleanup before proceeding with tests.

fabioh8010 commented:
**Updates:**

- Merged [JSON_PATCH solution branch](https://github.com/Expensify/react-native-onyx/compare/main...callstack-internal:react-native-onyx:bugfix/615-solution2-jsonPatchReplace) into this PR.
- Did manual tests with Native and Web:
  - Unit tests are passing in E/App ✅
  - Solution works in Native (SQLiteProvider) ✅
  - Solution works in Web (IDBKeyValProvider) ✅
- Did a couple of minor TS, tests and logic fixes.

Seems that Reassure is now reporting some performance degradation, I will investigate it further.

```
🔴 OnyxCache merge one call merging 10k keys > Duration deviation of 40.12 ms (40.74%) exceeded the allowed range of 10.00 ms (20.00%).
🔴 Onyx merge 10k calls with heavy objects > Duration deviation of 152.49 ms (35.04%) exceeded the allowed range of 10.00 ms (20.00%).
🔴 OnyxUtils initializeWithDefaultKeyStates one call initializing 10k heavy objects > Duration deviation of 108.72 ms (33.77%) exceeded the allowed range of 10.00 ms (20.00%).
🔴 Onyx init one call with 10k records to init > Duration deviation of 77.55 ms (33.54%) exceeded the allowed range of 10.00 ms (20.00%).
🔴 Onyx mergeCollection one call with 10k heavy objects > Duration deviation of 75.94 ms (28.81%) exceeded the allowed range of 10.00 ms (20.00%).
🔴 Onyx set 10k calls with heavy objects > Duration deviation of 65.38 ms (26.47%) exceeded the allowed range of 10.00 ms (20.00%).
🔴 Onyx multiSet one call with 10k heavy objects > Duration deviation of 49.66 ms (26.11%) exceeded the allowed range of 10.00 ms (20.00%).
🔴 Onyx update one call with 5k sets and 5k merges updates > Duration deviation of 91.49 ms (23.49%) exceeded the allowed range of 10.00 ms (20.00%).
🟢 OnyxUtils subscribeToKey one call subscribing to a whole collection of 10k heavy objects > Duration deviation of 25.56 ms (18.27%) is within the allowed range of 10.00 ms (20.00%).
🟢 Onyx setCollection one call with 10k heavy objects > Duration deviation of 36.80 ms (17.42%) is within the allowed range of 10.00 ms (20.00%).
🟢 OnyxUtils multiGet one call getting 10k heavy objects from storage > Duration deviation of 18.30 ms (13.81%) is within the allowed range of 10.00 ms (20.00%).
```

neil-marcellini commented:
Those performance hits don't look too bad to me! Maybe we have done enough optimization? It might be good to see how this works in App on some large accounts.

fabioh8010 commented:
**Updates:**

- Fixed `mergeCollection` not applying the solution in Web platform, and added a unit test to cover it.
- Add additional Storage checks in the unit tests covering this solution, to make sure both data in Storage and cache are the same.
- For next and final steps, add the remaining comments and open PR to review.

Note: I will be OOO tomorrow and Friday due to holidays, but I'll be back on Monday to focus on this.

chrispader commented on jest.config.js:
```diff
@@ -14,4 +14,8 @@ module.exports = {
     transformIgnorePatterns: ['node_modules/(?!((@)?react-native|@ngneat/falso|uuid)/)'],
+    moduleNameMapper: {
+        // Redirect all imports of OnyxMerge to its web version during unit tests.
+        '^(.*)/OnyxMerge$': '<rootDir>/lib/OnyxMerge/index.ts',
```
Can we somehow generalize this so this applies to all files with `.ts` or `.web.ts` file ending, so we don't have to update this in the future when we add more platform-specific files?

fabioh8010 replied:
I tried some approaches but I wasn't able yet to come up with something generic enough to cover all possible cases (also tried a custom resolver). I think we can explore this in a follow-up if you don't mind.

chrispader commented on lib/OnyxUtils.ts:
```diff
@@ -1142,77 +1145,101 @@ function hasPendingMergeForKey(key: OnyxKey): boolean {
+ * @param changes Array of changes that should be merged
+ * @param existingValue The existing value that should be merged with the changes
  */
 function applyMerge<TValue extends OnyxInput<OnyxKey> | undefined, TChange extends OnyxInput<OnyxKey> | undefined>(
```
To avoid confusion, could we rename this (internal) function to sth different, since we also have `OnyxMerge.applyMerge`.

This was my mistake when i refactored this part but i think it would be good to have unique function names for further discussion. Maybe `mergeImplementation` or `mergeInternal` is appropriate, since we are not exposing this function anyway

chrispader commented on lib/utils.ts:
```diff
@@ -35,83 +94,112 @@ function mergeObject<TObject extends Record<string, unknown>>(target: TObject |
+            return;
+        }
+
+        // If source value is not a mergable object, we need to set the source value it directly.
```
```suggestion
        // If the source value is not a mergable object, we need to set the key directly.
```

chrispader reviewed (dismissed):
Changes look good to me aside from a few minor requests 

Thanks for your great work @fabioh8010 ! 🙌🏼

ikevin127 commented:
**♻️ Testing the PR on the E/App, will drop a report soon.**

ikevin127 commented:
### 🧪 Manual Review Report

#### Report Actions (these directly use Onyx's merge/update )
- ✅ 
- - ✅ 
- - ✅ 
- - ✅ 
•  src/libs/actions/Report.ts
•  Test creation, updates, deletions, and interactions with report actions since these directly use Onyx's merge/update functionalities.
2. Expense Tracking and Submission:
•  src/pages/iou/request/step/IOURequestStepCategory.tsx
•  src/libs/actions/IOU.ts
•  Test tracking and submitting expenses following the steps in the issue, focusing on merging action updates related to expenses.
3. Pending Actions:
•  src/libs/actions/Transaction.ts
•  This area handles transaction actions, including nullifying pending actions, which are significant for the issue at hand.
4. User Interactions:
•  src/pages/home/ReportScreen.tsx
•  src/pages/home/report/ReportActionsView.tsx
•  Ensure user interactions such as marking actions as read or unread work correctly with updates to the Onyx cache.

Testing Focus

•  Interactions involving pendingAction: Ensure any functionality that marks actions as complete (nulls the pendingAction) operates correctly.
•  Race Conditions: Especially around clearing cache and restarting actions, which will heavily utilize the merge logic.
•  Consistency Across State Resets: Verify state consistency when cache is cleared, as this mimics real-world scenarios described in the issue.

ikevin127 commented:
- ✅

ikevin127 commented:
### 🧪 E/App Manual Review Report

> [!note]
> - performed the testing by building this PR's `react-native-onyx` and replacing the built `dist` in E/App
> - tests will be performed on all devices: **Android & iOS: HybridApp (Native), Android & iOS: mWeb, Web and Desktop**

#### SignUp / Login Flow
- ✅ on clear cache, sign up with a new account (public & private domain)
- ✅ while logged-in, logout and sign up with a new account (public & private domain)
- ✅ on clear cache, login with an existing account (public & private domain)
- ✅ while logged-in, login with another existing account (public & private domain)
- ✅ while anonymous in a public room, login / sign-up
  - once logged-in / signed-up -> navigate through the app to verify everything looks as expected
  
#### Troubleshoot: Clear Cache and Restart Functionality
- ✅ on a new account, start / create / add some messages in a: selfDM, 1-1 DM, workspace chat, group & room chats
  - verify that upon `Clear cache and restart` trigger, the LHN report previews look as expected for each of the 'start / create / add messages' report types

#### Report Actions (aka Messages / Comments: these directly use Onyx's merge/update)
- ✅ posting a message on a report / report thread (including offline -> online transition)
- ✅ updating a message on a report / report thread (including offline -> online transition)
- ✅ deleting a message on a report / report thread (including offline -> online transition)
- ✅ adding emojis on messages / nested thread messages (including offline -> online transition)

#### Expense Tracking and Submission
•  src/pages/iou/request/step/IOURequestStepCategory.tsx
•  src/libs/actions/IOU.ts
•  Test tracking and submitting expenses following the steps in the issue, focusing on merging action updates related to expenses.
- * [ ] TO BE CONTINUED...

#### Pending Actions
•  src/libs/actions/Transaction.ts
•  This area handles transaction actions, including nullifying pending actions, which are significant for the issue at hand.
- * [ ] TO BE CONTINUED...

#### User Interactions
•  src/pages/home/ReportScreen.tsx
•  src/pages/home/report/ReportActionsView.tsx
•  Ensure user interactions such as marking actions as read or unread work correctly with updates to the Onyx cache.
- * [ ] TO BE CONTINUED...

### ♻️ Testing focus

•  Interactions involving pendingAction: Ensure any functionality that marks actions as complete (nulls the pendingAction) operates correctly.
•  Race Conditions: Especially around clearing cache and restarting actions, which will heavily utilize the merge logic.
•  Consistency Across State Resets: Verify state consistency when cache is cleared, as this mimics real-world scenarios described in the issue.

ikevin127 reviewed (approved):
### 🧪 E/App Manual Review Report

> [!note]
> - performed the testing by building this PR's `react-native-onyx` and replacing the built `dist` in E/App
> - tests will be performed on all devices: **Android & iOS: HybridApp (Native), Android & iOS: mWeb, Web and Desktop**

#### SignUp / Login Flow
- ✅ on clear cache, sign up with a new account (public & private domain)
- ✅ while logged-in, logout and sign up with a new account (public & private domain)
- ✅ on clear cache, login with an existing account (public & private domain)
- ✅ while logged-in, login with another existing account (public & private domain)
- ✅ while anonymous in a public room, login / sign-up
- ✅ switch between ND / OD and back
  - once logged-in / signed-up -> navigate through the app to verify everything looks as expected
  
#### Troubleshoot: Clear Cache and Restart Functionality
- ✅ on a new account, start / create / add some messages in a: selfDM, 1-1 DM, workspace chat, group & room chats
  - verify that upon `Clear cache and restart` trigger, the LHN report previews look as expected for each of the 'start / create / add messages' report types

#### Report Actions (aka Messages / Comments: these directly use Onyx's merge/update)
- ✅ post / update / delete a message on a report / report thread (including offline -> online transition)
- ✅ add emojis on messages / nested thread messages (including offline -> online transition)
- ✅ search router functionality (`CMD + K`) outside of Reports page
- ✅ reports page functionality (filtering, sorting, manually typed queries)

#### Expense Tracking and Submission
- ✅ submit manual / scan / distance expense (including offline -> online transition)
- ✅ replace scan expense receipt after it was posted (including offline -> online transition)

#### Pending Actions (offline -> online transitions)
- ✅ while offline, post / update / delete a message on a report / report thread
- ✅ while offline, submit manual / scan / distance expense
  - then delete them, then return online

#### User Interactions
- ✅ mark reports read / unread (including offline -> online transition)
- ✅ pin / unpin reports (including offline -> online transition)
- ✅ expenses: approve / submit / mark as paid (including offline -> online transition)
- ✅ copy onyx / export onyx data functionality
- ✅ offline mode functionality

## Summary

🎉 No issues found, **E/App** behaving as expected with the update in all scenarios mentioned above.

neil-marcellini reviewed (commented):
Here's a very small start on reviewing this. I will tackle it tomorrow.

neil-marcellini commented on API.md:
```diff
@@ -157,7 +157,7 @@ Onyx.merge(ONYXKEYS.POLICY, {name: 'My Workspace'}); // -> {id: 1, name: 'My Wor
 <a name="mergeCollection"></a>
 
-## mergeCollection(collectionKey, collection)
+## mergeCollection(collectionKey, collection, mergeReplaceNullPatches)
```
Should we really be exposing the merge replacement patches parameter to the external facing API? It seems like something we are really only planning to use within Onyx. And I can't think of why Onyx users would need this.

I see that we do pass this value to merge collection within onyx. I think we should make a private and internal version of the function to keep the external facing API the same.

neil-marcellini commented on API-INTERNAL.md:
```diff
@@ -88,20 +82,6 @@ or if the provided key is a collection member key (in case our configured key is
 <dd><p>Tries to get a value from the cache. If the value is not present in cache it will return the default value or undefined.
 If the requested key is a collection, it will return an object with all the collection members.</p>
 </dd>
-<dt><a href="#removeLastAccessedKey">removeLastAccessedKey()</a></dt>
-<dd><p>Remove a key from the recently accessed key list.</p>
-</dd>
-<dt><a href="#addLastAccessedKey">addLastAccessedKey()</a></dt>
-<dd><p>Add a key to the list of recently accessed keys. The least
-recently accessed key should be at the head and the most
-recently accessed key at the tail.</p>
-</dd>
-<dt><a href="#addEvictableKeysToRecentlyAccessedList">addEvictableKeysToRecentlyAccessedList()</a></dt>
-<dd><p>Take all the keys that are safe to evict and add them to
-the recently accessed list when initializing the app. This
-enables keys that have not recently been accessed to be
-removed.</p>
-</dd>
```
Similar to another comment, why are we removing these?

fabioh8010 replied:
Leftovers from another PR, these methods are inside OnyxCache now.

neil-marcellini commented on tests/perf-test/OnyxUtils.perf-test.ts:
```diff
@@ -634,9 +625,7 @@ describe('OnyxUtils', () => {
-            await measureFunction(() =>
-                OnyxUtils.applyMerge(reportAction, [changedReportAction1, changedReportAction2, changedReportAction3, changedReportAction4, changedReportAction5], false),
-            );
+            await measureFunction(() => OnyxUtils.mergeChanges([changedReportAction1, changedReportAction2, changedReportAction3, changedReportAction4, changedReportAction5], reportAction));
```
Please rename the test case to `mergeChanges` from `applyMerge`.

neil-marcellini commented on tests/unit/onyxUtilsTest.ts:
I really like the way you wrote the tests in this file. I find them to be easy to read and track how the values are changing.

fabioh8010 replied:
Thanks!

neil-marcellini reviewed (changes requested):
I have reviewed the whole thing before you extracted the removeNestNulls thing, so I can go and quickly review that separate PR. I will also copy over any relevant comments.

I'm requesting changes because I have a few important questions and some small tweaks I would like to see. Great job on this and especially persisting with it over time.

Also as a heads up, I'm going to be out of office from July 4-16. If we don't get this merged by then for any reason, please go ahead without my review.

fabioh8010 commented on lib/OnyxMerge/index.native.ts:
```diff
@@ -0,0 +1,46 @@
+import type {ApplyMerge} from './types';
+
+const applyMerge: ApplyMerge = <TKey extends OnyxKey>(key: TKey, existingValue: OnyxValue<TKey>, validChanges: unknown[]) => {
+    // If any of the changes is null, we need to discard the existing value.
+    const baseValue = validChanges.includes(null) ? undefined : existingValue;
```
When one of the `validChanges` are `null`, it means that we actually want to fully replace that value in Storage (like a SET). If we are going to fully replace the existing value we don't need it anymore, so we just merge all the `validChanges` that, together, will become the new value.

neil-marcellini replied:
Ah I see, it makes sense that `null` is included in `validChanges` to signal that the value should be replaced.

fabioh8010 commented on lib/Onyx.ts:
```diff
@@ -718,38 +686,48 @@ function update(data: OnyxUpdate[]): Promise<void> {
+            return;
         }
+
+        const mergePromises = operations.map((operation) => {
+            return merge(key, operation);
+        });
+        promises.push(() => mergePromises.at(0) ?? Promise.resolve());
```
I consulted @chrispader about this one

neil-marcellini replied:
This can be NAB because the tests and app seem to work well. I will await Chris' answer because I'm still curious.

fabioh8010 commented on lib/storage/providers/SQLiteProvider.ts:
```diff
@@ -82,57 +103,78 @@ const provider: StorageProvider = {
 
-        return this.multiMerge([[key, deltaChanges]]) as Promise<BatchQueryResult>;
+        commands.push({query: patchQuery, params: patchQueryArguments});
+        if (replaceQueryArguments.length > 0) {
+            commands.push({query: replaceQuery, params: replaceQueryArguments});
+        }
```
Yeah I think we can explore this in a follow up PR, I will also include [this](https://github.com/Expensify/react-native-onyx/pull/620#discussion_r2179286146) if you don't mind cc @lakchote

fabioh8010 commented:
@neil-marcellini Addressed/asked all comments!

neil-marcellini reviewed (approved):
Thank you for the updates. It all looks good to me now. I think it's good to go. We can follow up on some NAB items later.
